python/tk_nuke/menu_generation.py

```diff
@@ -21,291 +18,665 @@
+        """
+        ctx = self.engine.context
+
+        if ctx.entity is None:
+            ctx_name = "%s" % ctx.project["name"]
+        elif ctx.step is None and ctx.task is None:
+            # entity only
+            # e.g. Shot ABC_123
+            ctx_name = "%s %s" % (ctx.entity["type"], ctx.entity["name"])
+        else:
+            # we have either step or task
+            task_step = None
+            if ctx.step:
+                task_step = ctx.step.get("name")
+            if ctx.task:
+                task_step = ctx.task.get("name")
+
+            # e.g. [Lighting, Shot ABC_123]
+            ctx_name = "%s, %s %s" % (task_step, ctx.entity["type"], ctx.entity["name"])
+
+        # create the menu object
+        ctx_menu = self._menu_handle.addMenu(ctx_name)
+        action = ctx_menu.addAction("Jump to Shotgun")
+        action.triggered.connect(self._jump_to_sg)
+        action = ctx_menu.addAction("Jump to File System")
+        action.triggered.connect(self._jump_to_fs)
+        ctx_menu.addSeparator()
+        return ctx_menu
+
+    def _add_app_menu(self, commands_by_app):
+        """
+        Add all apps to the main menu.
+        """
+        for app_name in sorted(commands_by_app.keys()):
+            if len(commands_by_app[app_name]) > 1:
+                # more than one menu entry fort his app
+                # make a sub menu and put all items in the sub menu
+                app_menu = self._menu_handle.addMenu(app_name)
+                for cmd in commands_by_app[app_name]:
+                    cmd.add_command_to_menu(app_menu)
+            else:
+                # this app only has a single entry.
+                # display that on the menu
+                # todo: Should this be labeled with the name of the app
+                # or the name of the menu item? Not sure.
+                cmd_obj = commands_by_app[app_name][0]
+                if not cmd_obj.favourite:
+                    # skip favourites since they are already on the menu
+                    cmd_obj.add_command_to_menu(self._menu_handle)
+
+# -----------------------------------------------------------------------------
+
+class NukeMenuGenerator(MenuGenerator):
+    """
+    A Nuke specific menu generator.
+    """
+    def __init__(self, *args, **kwargs):
+        super(NukeMenuGenerator, self).__init__(*args, **kwargs)
+        self._dialogs = []
+        engine_root_dir = self.engine.disk_location
+        self._shotgun_logo = os.path.abspath(
+            os.path.join(
+                engine_root_dir,
+                "resources",
+                "sg_logo_80px.png",
+            ),
+        )
+        self._shotgun_logo_blue = os.path.abspath(
+            os.path.join(
+                engine_root_dir,
+                "resources",
+                "sg_logo_blue_32px.png",
+            ),
+        )
+
+    def create_menu(self):
+        """
+        Creates the "Shotgun" menu in Nuke.
```

Comment:
I should note, however, that the Nuke menu creation logic will work just fine when we get to Nuke Studio, so we shouldn't need a third implementation there.